### ISSUE TITLE
Infer Alliance Selection From QF Schedule

### ISSUE THREAD

**Issue by phil-lopreiato** [the-blue-alliance/the-blue-alliance#1456]
The FRCAPI Alliances endpoint doesn't get populated until after the first qf match has been played, so we currently don't send Alliance Selection notifications/display the data in-app until then. 

[From Alex in TeamForge](https://usfirst.collab.net/sf/go/artf4844):

> Currently, the alliances are only updated at the end of a match, which is why the Quarterfinal schedule doesn't correlate with alliances. They 'appear as the event goes on' because after a match is played, they're uploaded. We can explore changing this in the future, but at the moment, it's behaving as expected. I've opened an internal feature request, instead.

We can instead use the first 4 quarterfinal matches plus team player station positions to know what the initial alliance selections were. The captain will always be in station 2, pick 1 in station 1, and pick 2 in station 3. 

From 2016 Manual, Section 5:

> ALLIANCE Leads will always be assigned to the center PLAYER STATION, the first pick will be assigned to the station to their left while they’re facing the FIELD, and the second pick will be assigned to their right while they’re facing the FIELD. If a BACKUP TEAM is in play, they will be assigned to the PLAYER STATION that was assigned to the DRIVE TEAM they’re replacing.

![bracket](https://cloud.githubusercontent.com/assets/2754863/14233698/d9c655ae-f99d-11e5-8cc7-3460dcf7c401.png)

EDIT: Might want to hold off on this for now, Alex updated the tracker with the following:

> I expect to have more info, or a change, for Week 6.

**Comment by 1fish2** [comment]
If Alex needs more justification for fixing this, point out that after alliance selections there's usually lunch break and other stuff before QF matches. During that time, teams need to plan their strategy (and this year, select tower defenses) depending on the specific alliances and team capabilities.

**Comment by gregmarra** [comment]
It was pointed out to me at Boston today how silly this gap is, given we
know the other schedules. :)

This is a good one to have unit tests for, and probably some test using
live FRCFMS data (do we have tests that do that?) to confirm the format
never changes. Would be bad to mis-report alliances. Alliances changing in
production should probably throw an error.

**Comment by phil-lopreiato** [comment]
I think the plan should first be to wait and see if FIRST is going to make any changes to the behavior of the alliances endpoint for week 6. If they do change the behavior to post data as soon as the scorekeeper posts the schedule, then we don't need to do anything. If they decide to maintain the current behavior for now, then I agree we should figure out the alliances ourselves.

A couple edge cases we will have to worry about and reasons why I think the endpoint (whenever it gets posted) should be able to override schedule-based guesses:
 - backup teams (I bet we could break it by calling in a backup team at a "bad" time)
 - data entry error (alliance selection is the most difficult part of the event for a scorekeeper, it's possible that data entry errors could exist and have to be corrected)
 - 4 team alliances at CMP (the schedule would only provide for 3/4, but it won't specify which 3)

**Comment by phil-lopreiato** [comment]
According to the [Octofinal Test Data](https://gist.github.com/phil-lopreiato/b498f2a3ab9ae4681ffe4274db6d3d36) (see #1436), here's the alliance mapping (red v blue):
 - EF1: 1 v 16
 - EF2: 8 v 9
 - EF3: 4 v 13
 - EF4: 5 v 12
 - EF5: 2 v 15
 - EF6: 7 v 10
 - EF7: 3 v 14
 - EF8: 6 v 11

**Comment by 1fish2** [comment]
It seems that FIRST did not fix it for week 6, given that the Android app showed the QF schedule for SVR long before showing the alliance list.

**Comment by phil-lopreiato** [comment]
I do not believe there were any changes for last week. @aherreidFIRST (or would you rather communicate on TeamForge for things like this?) do you plan on a change for Week 7 or CMP?

**Comment by phil-lopreiato** [comment]
This should be fixed for Week 7 and beyond, so we can back-burner this for now.

However, I still think we can do it for offseason events where don't have alliance data directly input.

**Comment by aherreidFIRST** [comment]
Just to add some details, so you know:
> Data entry errors - Obviously we can't control teams being mis-entered. But, FMS validates that the team exists at the event (of course) so it's at least going to be from the event (if you use the registration list)
> 4 team alliances - As the lineups are submitted, the schedule will change to reflect who was selected. But, in terms of initial placement, it works the same as traditional 8 alliance events. AKA, the third round pick won't appear in the schedule at all until subbed in via a lineup.
Hope that helps.

**Comment by phil-lopreiato** [comment]
Thanks for the additional details!

> But, in terms of initial placement, it works the same as traditional 8 alliance events. AKA, the third round pick won't appear in the schedule at all until subbed in via a lineup.

This is good to know (and would have been my next question when I started implementing :smile:)
